File: leetcode2181/solution.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* mergeNodes(ListNode* head) {
        /*
        Time: O(n) | Space: O(1)

        where 'n' is the number of nodes in the given list.
        */
        ListNode* left = head;
        ListNode* right = head->next;

        while (left != NULL && right != NULL) {
            if (right->val == 0) {
                if (right->next == NULL) {
                    left->next = NULL;
                    break;
                }
                left->next = right;
                left = left->next;
            } else {
                left->val += right->val;
            }
            right = right->next;
        }


        return head;
    }
};

File: leetcode2181/solution.rs
// Definition for singly-linked list.
// #[derive(PartialEq, Eq, Clone, Debug)]
// pub struct ListNode {
//   pub val: i32,
//   pub next: Option<Box<ListNode>>
// }
// 
// impl ListNode {
//   #[inline]
//   fn new(val: i32) -> Self {
//     ListNode {
//       next: None,
//       val
//     }
//   }
// }

impl Solution {
    pub fn merge_nodes(head: Option<Box<ListNode>>) -> Option<Box<ListNode>> {
        /// Time: O(n) | Space: O(1)
        /// 
        /// where n is the number of nodes in the list
        let mut res: Option<Box<ListNode>> = None;
        let mut left: &mut Option<Box<ListNode>> = &mut res;
        let mut right: Option<Box<ListNode>> = head;

        while let Some(mut node) = right {
            let mut sum = node.val;

            loop {
                node = node.next.unwrap();

                if node.val == 0 {
                    right = node.next.take();
                    node.val = sum;
                    left = &mut left.insert(node).next;
                    break;
                }

                sum += node.val;
            }
        }

        res
        
    }
}